sqlccl: add naive progress tracking to backup and restore

benesch opened the pull request:
See commit messages for details. @danhhz, I wanted to get your approval on the `ErrGroupWithProgress` before I wrote tests for it, but this PR already sports tests for `JobLogger.Progressed()` and backup/restore incremental progress as a whole.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14459)
<!-- Reviewable:end -->

benesch committed:
sql: teach JobLogger to track the progress of jobs

This commit adds a JobLogger.Progressed() method that updates the
fraction_completed field on a job in system.jobs.

benesch committed:
sqlccl: add naive progress tracking to backup and restore

This commit adds a goroutine that updates the FractionCompleted field in
the system.jobs table every second during backups and restores. The
computation of FractionCompleted is rather simple in this first
implementation: we chunk the requests for backup into range-sized pieces
(this already happens for restore), then divide the number of completed
requests by the total number of requests.

@danhhz and I expect this computation to be more reasonable than it may
at first seem. The error is highest in small backups, where ranges are
likely to be vastly different sizes, but the demand for accurate
progress when the job is quick is low. In larger backups, where accurate
progress is important, we expect the computation to be more accurate,
since most ranges will be between 32 and 64MB in size.

danhhz commented:
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup.go, line 193 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgV6rzYegtMrrFAwwYX:-KgV6rzYegtMrrFAwwYY:b9ttzek) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68696fd555bde73e3086d7aa3e93737cc111813/pkg/ccl/sqlccl/backup.go#L193)):*
> ```Go
> 		for _, payload := range split.Payload.([]interface{}) {
> 			if _, ok := payload.(spanMarker); ok {
> 				needed = true
> ```

nit: you can break here

---

*[pkg/ccl/sqlccl/backup.go, line 377 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgVbkmBcywk3BRCwTEn:-KgVbkmBcywk3BRCwTEo:b-ivl8ss) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68696fd555bde73e3086d7aa3e93737cc111813/pkg/ccl/sqlccl/backup.go#L377)):*
> ```Go
> 		fractionCompleted := float32(mu.responseCount) / float32(len(spans))
> 		mu.Unlock()
> 		if err := jobLogger.Progressed(ctx, fractionCompleted); err != nil {
> ```

let's just update when a request comes back (modulo some rate limiting). it's considerably simpler and it's not like this number will change if requests aren't coming back

---

*[pkg/ccl/sqlccl/backup_test.go, line 340 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgVc3wOMV6cS15PaWcS:-KgVc3wOMV6cS15PaWcT:bhnjcmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68696fd555bde73e3086d7aa3e93737cc111813/pkg/ccl/sqlccl/backup_test.go#L340)):*
> ```Go
> 	var mu syncutil.Mutex
> 	params := base.TestClusterArgs{}
> 	params.ServerArgs.Knobs.Store = &storage.StoreTestingKnobs{
> ```

we know how many requests this will get. (that number might change if the data splits, but it shouldn't the timeline of this test). how about a filter that lets the first ~50% of requests finish and blocks everything after that on a sync.Cond. then you can check that the progress gets updated to between 25% and 75%, trigger the cond, and let the test finish

feels less likely to flake to me (and you could potential get rid of the ProgressReportingInterval mutation below, which I'd prefer to avoid if we can)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/backup.go, line 193 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgV6rzYegtMrrFAwwYX:-KgW9kft2jLhd9aMIFsX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68696fd555bde73e3086d7aa3e93737cc111813/pkg/ccl/sqlccl/backup.go#L193)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: you can break here
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.go, line 377 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgVbkmBcywk3BRCwTEn:-KgW9lsLbCX-1LYsVD4U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68696fd555bde73e3086d7aa3e93737cc111813/pkg/ccl/sqlccl/backup.go#L377)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

let's just update when a request comes back (modulo some rate limiting). it's considerably simpler and it's not like this number will change if requests aren't coming back
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup_test.go, line 340 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgVc3wOMV6cS15PaWcS:-KgW9nXN7rwXgY1R1ceh:b-9lmk1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68696fd555bde73e3086d7aa3e93737cc111813/pkg/ccl/sqlccl/backup_test.go#L340)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

we know how many requests this will get. (that number might change if the data splits, but it shouldn't the timeline of this test). how about a filter that lets the first ~50% of requests finish and blocks everything after that on a sync.Cond. then you can check that the progress gets updated to between 25% and 75%, trigger the cond, and let the test finish

feels less likely to flake to me (and you could potential get rid of the ProgressReportingInterval mutation below, which I'd prefer to avoid if we can)
</blockquote></details>

It ended up being mostly as simple to ensure that every response triggers a progress update, so I went with that approach. The still-present complexity stems from needing to spawn a separate goroutine to check the updated progress partway through the job (since `sqlDB.Exec("BACKUP")` is, understandably, blocking).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: 0 of 5 files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/backup.go, line 216 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgWU0wwKgM6OVoEG65C:-KgWU0wwKgM6OVoEG65D:bh5ygnq) ([raw file](https://github.com/cockroachdb/cockroach/blob/92e33fd7e2e1db5ac28aa93b967098328702a70c/pkg/ccl/sqlccl/backup.go#L216)):*
> ```Go
> 				EndKey: roachpb.Key(split.End),
> 			})
> 			fmt.Printf("backup %+v\n", splitSpans[len(splitSpans)-1])
> ```

debugging?

---

*[pkg/ccl/sqlccl/backup.go, line 331 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZimrUDdTGGuez6OHR:-KgZimrUDdTGGuez6OHS:bv6e41e) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup.go#L331)):*
> ```Go
> 	if err := db.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
> 		var err error
> 		ranges, err = allRangeDescriptors(ctx, txn)
> ```

put a TODO that we could fetch only the ranges we need (which will be important when backing up small fractions of very large clusters)

---

*[pkg/ccl/sqlccl/backup.go, line 338 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZj2dSRl_3KX2UJU9m:-KgZj2dSRl_3KX2UJU9n:b-hdwlk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup.go#L338)):*
> ```Go
> 
> 	// We split the spans into range-sized pieces so that no request is
> 	// responsible for more than 64MB of data. This allows us to use the number of
> ```

very nitpicky nit: ranges can be more than 64MB when the split queue gets backed up. "We split the spans into range-sized pieces so that we can use the number of completed requests as a rough measure of progress." is just as informative and slightly more correct

---

*[pkg/ccl/sqlccl/backup.go, line 390 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZjp1ce2w-78kbZELX:-KgZjp1dqGwYwA0oq2b_:b-antntj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup.go#L390)):*
> ```Go
> 				mu.lastReportedAt = timeutil.Now()
> 				mu.lastReportedFraction = fractionCompleted
> 				if err := jobLogger.Progressed(ctx, fractionCompleted); err != nil {
> ```

comment why this error is logged and not returned

---

*[pkg/ccl/sqlccl/backup.go, line 391 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZjxlgVpsUU5fuzBE9:-KgZjxlgVpsUU5fuzBEA:bwdzlp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup.go#L391)):*
> ```Go
> 				mu.lastReportedFraction = fractionCompleted
> 				if err := jobLogger.Progressed(ctx, fractionCompleted); err != nil {
> 					log.Errorf(ctx, "BACKUP ignoring error while updating progress on job '%s' to %f: %+v",
> ```

include the jobid in this error message, plz

---

*[pkg/ccl/sqlccl/backup_test.go, line 340 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgVc3wOMV6cS15PaWcS:-KgZs20AJvIYFzGDUiUR:b-kmbou8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68696fd555bde73e3086d7aa3e93737cc111813/pkg/ccl/sqlccl/backup_test.go#L340)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

It ended up being mostly as simple to ensure that every response triggers a progress update, so I went with that approach. The still-present complexity stems from needing to spawn a separate goroutine to check the updated progress partway through the job (since `sqlDB.Exec("BACKUP")` is, understandably, blocking).
</blockquote></details>

I think this is simpler if you invert which thing is run asynchronously. plus you can get rid of the util.RetryForDuration which should always be avoided if it can. i might be missing something but maybe something like the following

```
errChan := make(chan error)
go func() { _, err := sqlDB.DB.Exec("BACKUP ..."); errChan <- err }
defer func() {
   // Make sure we don't leak any goroutines
   if err := <-errChan; err != nil {
     ...
   }
}
// We are guaranteed at least backupRestoreDefaultRanges requests
for i := 0; i < backupRestoreDefaultRanges; i++ {
  _ <-responseComplete
  var fractionCompleted float32
  sqlDB.QueryRow(...).Scan(&fractionCompleted)
  if fractionCompleted is not in some range {
    t.Errorf()
  }
}
```

---

*[pkg/ccl/sqlccl/backup_test.go, line 108 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZuIZYIkClFi-rMvQI:-KgZuIZYIkClFi-rMvQJ:b-daqhtj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup_test.go#L108)):*
> ```Go
> }
> 
> func backupRestoreTestSetupWithParams(
> ```

I don't think this helper gets you anything, just inline the bits you need (basically what was there before but with TestCluster instead of TestServer)

---

*[pkg/ccl/sqlccl/backup_test.go, line 287 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZrLlvqk-TEpovUb2O:-KgZrLlvqk-TEpovUb2P:bf6d0k1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup_test.go#L287)):*
> ```Go
> }
> 
> func startTrackingSystemJobProgress(db *gosql.DB, responseComplete chan bool) func(int) error {
> ```

this deserves a function level comment, especially what the return value is

---

*[pkg/ccl/sqlccl/backup_test.go, line 333 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZsANiHf7ydtdTYkC_:-KgZsANiHf7ydtdTYkCa:benwcji) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup_test.go#L333)):*
> ```Go
> 		if e, a := minDistinct, len(seenFractions); a < e {
> 			return errors.Errorf(
> 				"expected job progress to have at least %d distinct immediate fractions, but got only %d: %+v",
> ```

intermediate

---

*[pkg/ccl/sqlccl/backup_test.go, line 355 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZulJ11W-Idokn0dZp:-KgZulJ2YyruRiXGuMtr:blo7h28) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup_test.go#L355)):*
> ```Go
> 	// the progress is updated in the system.jobs table before it does its next
> 	// receive, thus ensuring that each response causes a progress update.
> 	responseComplete := make(chan bool)
> ```

chan struct{}

---

*[pkg/ccl/sqlccl/backup_test.go, line 375 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZsEdV-oOP7PRj9HNx:-KgZsEdV-oOP7PRj9HNy:b-moo225) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup_test.go#L375)):*
> ```Go
> 	const numAccounts = 1000
> 
> 	// We require that we see a progress update for at least each range that we've
> ```

this comment could be clarified to make it more obvious that this is checking preconditions of the test, not checking the impl

---

*[pkg/ccl/sqlccl/restore.go, line 442 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgWUIiPm8FO-9_kyHkI:-KgWUIiQuCF8O0JbZHar:bh5ygnq) ([raw file](https://github.com/cockroachdb/cockroach/blob/92e33fd7e2e1db5ac28aa93b967098328702a70c/pkg/ccl/sqlccl/restore.go#L442)):*
> ```Go
> 				files:     files,
> 			})
> 			fmt.Printf("request %+v\n", requestEntries[len(requestEntries)-1])
> ```

debugging?

---

*[pkg/ccl/sqlccl/restore.go, line 650 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZkHLP1-hrAm23P4Da:-KgZkHLP1-hrAm23P4Db:bw9ahoi) ([raw file](https://github.com/cockroachdb/cockroach/blob/34a5b40de9010f36f0a7bfda3ab21b19eb039820/pkg/ccl/sqlccl/restore.go#L650)):*
> ```Go
> 		ir := importRequests[i]
> 		g.Go(func() error {
> 			return Import(gCtx, db, ir.Key, ir.EndKey, ir.files, kr)
> ```

if err != nil, return it immediately. we don't need to update progress for a job we're about to fail (plus is seems incorrect to include failed requests in progress)

---

*[pkg/ccl/sqlccl/restore.go, line 672 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZkVMqTOcyYuB5sBW3:-KgZkVMr-caOiENaJgwa:bcdp0gj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/restore.go#L672)):*
> ```Go
> 				if err := jobLogger.Progressed(ctx, fractionCompleted); err != nil {
> 					log.Errorf(ctx, "RESTORE ignoring error while updating progress on job '%s' to %f: %+v",
> 						jobLogger.Job.Description, fractionCompleted, err)
> ```

ditto comment and jobid

---

*[pkg/sql/jobs_test.go, line 177 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZmbaTxDN8OwRkaFVF:-KgZmbaTxDN8OwRkaFVG:b-mzu90q) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e66161a4ec2ae4a7b4e292e78ec1d6bc716276/pkg/sql/jobs_test.go#L177)):*
> ```Go
> 		}
> 
> 		// This fraction completed progression is intentionally not strictly
> ```

it's fine to test the more general case here, but fwiw, I don't think the backup/restore tracking does this. individual calls may fail, but the locking serializes the requests

though now that you mention it, we probably shouldn't be sending the requests in the locked section

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459)*
<!-- Sent from Reviewable.io -->

benesch committed:
sqlccl: include job ID in job logging error messages

benesch commented:
Review status: 0 of 5 files reviewed at latest revision, 17 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/backup.go, line 216 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgWU0wwKgM6OVoEG65C:-Kg_V5wslOXgzjUg2zje:ble0ssl) ([raw file](https://github.com/cockroachdb/cockroach/blob/92e33fd7e2e1db5ac28aa93b967098328702a70c/pkg/ccl/sqlccl/backup.go#L216)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

debugging?
</blockquote></details>

Yup, already fixed by the revision after the one you reviewed.

---

*[pkg/ccl/sqlccl/backup.go, line 331 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZimrUDdTGGuez6OHR:-Kg_VZO17rt96gavQTc2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup.go#L331)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

put a TODO that we could fetch only the ranges we need (which will be important when backing up small fractions of very large clusters)
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.go, line 338 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZj2dSRl_3KX2UJU9m:-Kg_VSE9T5bfzZQl2dT8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup.go#L338)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

very nitpicky nit: ranges can be more than 64MB when the split queue gets backed up. "We split the spans into range-sized pieces so that we can use the number of completed requests as a rough measure of progress." is just as informative and slightly more correct
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.go, line 390 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZjp1ce2w-78kbZELX:-Kg_i5d0NBPvIEKxri2r:b-ispbq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup.go#L390)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

comment why this error is logged and not returned
</blockquote></details>

We don't do that in the `Succeeded` error message below. I'll take care of both of them in another commit.

---

*[pkg/ccl/sqlccl/backup.go, line 391 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZjxlgVpsUU5fuzBE9:-Kg_i6IlvvJCTPSuiKrz:bs4q9op) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup.go#L391)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

include the jobid in this error message, plz
</blockquote></details>

Ditto.

---

*[pkg/ccl/sqlccl/backup_test.go, line 340 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgVc3wOMV6cS15PaWcS:-Kg_VtsH2ZDsO9odkyXI:b-wkwk97) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68696fd555bde73e3086d7aa3e93737cc111813/pkg/ccl/sqlccl/backup_test.go#L340)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I think this is simpler if you invert which thing is run asynchronously. plus you can get rid of the util.RetryForDuration which should always be avoided if it can. i might be missing something but maybe something like the following

```
errChan := make(chan error)
go func() { _, err := sqlDB.DB.Exec("BACKUP ..."); errChan <- err }
defer func() {
   // Make sure we don't leak any goroutines
   if err := <-errChan; err != nil {
     ...
   }
}
// We are guaranteed at least backupRestoreDefaultRanges requests
for i := 0; i < backupRestoreDefaultRanges; i++ {
  _ <-responseComplete
  var fractionCompleted float32
  sqlDB.QueryRow(...).Scan(&fractionCompleted)
  if fractionCompleted is not in some range {
    t.Errorf()
  }
}
```


</blockquote></details>

You need the `util.RetryForDuration` because the store filter executes *before* `jobLogger.Progressed` is called. Unfortunate, I know. I'm open to other ideas. Using `RetryForDuration` is actually far less flaky than a strategy which verifies progress is at least `n - 5` on request `n`, for example.

I think you're right that flipping which bit goes in the goroutine makes the code a bit clearer. I've pushed a new version with that approach. It's still complicated, though, to handle a bunch of sync edge cases. Your sample will hang forever, for example, if a bug is introduced that causes backup or restore to issue less than `backupRestoreDefaultRanges` requests.

---

*[pkg/ccl/sqlccl/backup_test.go, line 108 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZuIZYIkClFi-rMvQI:-Kg_iNBYYu0G5RkFZcWp:bizrb91) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup_test.go#L108)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I don't think this helper gets you anything, just inline the bits you need (basically what was there before but with TestCluster instead of TestServer)
</blockquote></details>

I need data that's split across ten ranges, no? And at that point it's the entire function.

---

*[pkg/ccl/sqlccl/backup_test.go, line 287 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZrLlvqk-TEpovUb2O:-Kg_idPp4AGM8A6kOGV7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup_test.go#L287)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

this deserves a function level comment, especially what the return value is
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup_test.go, line 333 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZsANiHf7ydtdTYkC_:-Kg_idk0LbPeecXm1MvN:b-dtkdrq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup_test.go#L333)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

intermediate
</blockquote></details>

Done, thanks.

---

*[pkg/ccl/sqlccl/backup_test.go, line 355 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZulJ11W-Idokn0dZp:-Kg_n-WXzd4eK8-71n4W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup_test.go#L355)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

chan struct{}
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup_test.go, line 375 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZsEdV-oOP7PRj9HNx:-Kg_n-rKVTJhD0UxZcQ9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup_test.go#L375)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

this comment could be clarified to make it more obvious that this is checking preconditions of the test, not checking the impl
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/restore.go, line 442 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgWUIiPm8FO-9_kyHkI:-Kg_WBHWa0tOr-92DP4C:bp5kgyl) ([raw file](https://github.com/cockroachdb/cockroach/blob/92e33fd7e2e1db5ac28aa93b967098328702a70c/pkg/ccl/sqlccl/restore.go#L442)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

debugging?
</blockquote></details>

(Also already removed.)

---

*[pkg/ccl/sqlccl/restore.go, line 650 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZkHLP1-hrAm23P4Da:-Kg_Xr_WZlV1pqmMU906:b-qu6n94) ([raw file](https://github.com/cockroachdb/cockroach/blob/34a5b40de9010f36f0a7bfda3ab21b19eb039820/pkg/ccl/sqlccl/restore.go#L650)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

if err != nil, return it immediately. we don't need to update progress for a job we're about to fail (plus is seems incorrect to include failed requests in progress)
</blockquote></details>

Good point. Done.

---

*[pkg/ccl/sqlccl/restore.go, line 672 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZkVMqTOcyYuB5sBW3:-Kg_n0rrKZZy6Srf7L37:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/restore.go#L672)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

ditto comment and jobid
</blockquote></details>

Done.

---

*[pkg/sql/jobs_test.go, line 177 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZmbaTxDN8OwRkaFVF:-Kg_WSAbMq6_J7f9AVUt:bm30uz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e66161a4ec2ae4a7b4e292e78ec1d6bc716276/pkg/sql/jobs_test.go#L177)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

it's fine to test the more general case here, but fwiw, I don't think the backup/restore tracking does this. individual calls may fail, but the locking serializes the requests

though now that you mention it, we probably shouldn't be sending the requests in the locked section
</blockquote></details>

Updated the comment to be more clear that this doesn't currently happen.

Re the lock: I wanted to avoid having to call `mu.Unlock()` in separate code paths, and I figured it didn't matter because the waitgroup won't return until the final call to `Progressed` completes anyway, so who cares if the other threads queue up waiting for the mutex. In any case, I've pushed a version that calls `Unlock()` before calling `Progressed` for your review.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r1, 4 of 4 files at r3, 3 of 4 files at r7, 3 of 3 files at r8.
Review status: 4 of 5 files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup.go, line 59 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kg_ptq28Lu6Yl3yf-rC:-Kg_ptq28Lu6Yl3yf-rD:b1vl32i) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ebd9bd67caa2c7d64af35aa82ded0d6584ee94b/pkg/ccl/sqlccl/backup.go#L59)):*
> ```Go
> func shouldLogProgress(fraction float32, lastFraction float32, lastReported time.Time) bool {
> 	return lastReported.Add(progressTimeThreshold).Before(timeutil.Now()) ||
> 		fraction-lastFraction > progressCompletionThreshold
> ```

nit: check this first since it short-circuits

---

*[pkg/ccl/sqlccl/backup.go, line 388 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kg_q9ZRvPxO8nqch0XH:-Kg_q9ZRvPxO8nqch0XI:bmpvzbr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ebd9bd67caa2c7d64af35aa82ded0d6584ee94b/pkg/ccl/sqlccl/backup.go#L388)):*
> ```Go
> 			mu.responseCount++
> 			fractionCompleted := float32(mu.responseCount) / float32(len(spans))
> 			if shouldLogProgress(fractionCompleted, mu.lastReportedFraction, mu.lastReportedAt) {
> ```

can the duplication between this and restore be reduced? it looks like it stems from sharing `mu.Mutex`, which is different in each of the two files, but I don't see why updates to progress need to be atomic with updates to `mu.files`, say.

---

*[pkg/ccl/sqlccl/backup.go, line 510 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kg_r7MJFd05eW1fb6Eg:-Kg_r7MJFd05eW1fb6Eh:byu28y8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cbb9ad02ea8404c20d20ed11c2f87819b02e98/pkg/ccl/sqlccl/backup.go#L510)):*
> ```Go
> 		}
> 		if err := jobLogger.Succeeded(ctx); err != nil {
> 			// An error while marking the job as successful is not important enough to
> ```

belongs on the previous commit. squash the two?

---

*[pkg/ccl/sqlccl/backup_test.go, line 288 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kg_rO78dqNHBqF5COV7:-Kg_rO79eEDEJ-Kyi7tD:b-i74mt8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cbb9ad02ea8404c20d20ed11c2f87819b02e98/pkg/ccl/sqlccl/backup_test.go#L288)):*
> ```Go
> 
> // verifySystemJobProgress repeatedly asserts that a progress update occurs soon
> // after it receives a message on the responseComplete channel. (A progress
> ```

an entire sentence in parentheses?

---

*[pkg/sql/jobs_test.go, line 116 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kg_oMwUM_njyT_oOSAb:-Kg_oMwV5YckvOE0Lk_G:b-wu2l2w) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e66161a4ec2ae4a7b4e292e78ec1d6bc716276/pkg/sql/jobs_test.go#L116)):*
> ```Go
> 		)
> 	}
> 	if status == sql.JobStatusPending {
> ```

i think this might be clearer as a `switch status`?

---

*[pkg/sql/jobs_test.go, line 350 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kg_opHizBvCO2ICurmR:-Kg_opHizBvCO2ICurmS:b-j16y7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e66161a4ec2ae4a7b4e292e78ec1d6bc716276/pkg/sql/jobs_test.go#L350)):*
> ```Go
> 	})
> 
> 	t.Run("progress on finished job fails", func(t *testing.T) {
> ```

suggest you run this with `-v` before you settle on multiword subtest names

---

*[pkg/sql/jobs_test.go, line 178 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kg_p3O-l8NwlAhU2NR4:-Kg_p3O-l8NwlAhU2NR5:b-e9okx7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ebd9bd67caa2c7d64af35aa82ded0d6584ee94b/pkg/sql/jobs_test.go#L178)):*
> ```Go
> 
> 		// This fraction completed progression is intentionally not strictly
> 		// increasing to simulate imprecise progress estimates.
> ```

this belongs in the preceding commit

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 1 of 5 files reviewed at latest revision, 24 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/backup.go, line 390 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgZjp1ce2w-78kbZELX:-KgltvmCSRsaWiXWCTYO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8658321cba821f0417c5c16fe3a785b8df9da69/pkg/ccl/sqlccl/backup.go#L390)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

We don't do that in the `Succeeded` error message below. I'll take care of both of them in another commit.
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.go, line 59 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kg_ptq28Lu6Yl3yf-rC:-Kglav-ScpOF6QJkZc2V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ebd9bd67caa2c7d64af35aa82ded0d6584ee94b/pkg/ccl/sqlccl/backup.go#L59)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: check this first since it short-circuits
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.go, line 388 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kg_q9ZRvPxO8nqch0XH:-Kgltek7vmSRy8l3nM4r:b-2usgcj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ebd9bd67caa2c7d64af35aa82ded0d6584ee94b/pkg/ccl/sqlccl/backup.go#L388)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can the duplication between this and restore be reduced? it looks like it stems from sharing `mu.Mutex`, which is different in each of the two files, but I don't see why updates to progress need to be atomic with updates to `mu.files`, say.
</blockquote></details>

Those two updates don't need to be atomic, but having to acquire two mutexes when the job comes back grates a bit.

@danhhz, do you have thoughts? I want to know your thoughts on extracting a `BackupRestoreProgressMonitor` before I go down that road. I know you're in general less worried about duplicating code.

---

*[pkg/ccl/sqlccl/backup.go, line 510 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kg_r7MJFd05eW1fb6Eg:-KgaPoPqhvOJz6e7mv6X:b-j6zbd0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cbb9ad02ea8404c20d20ed11c2f87819b02e98/pkg/ccl/sqlccl/backup.go#L510)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

belongs on the previous commit. squash the two?
</blockquote></details>

This log statement's already merged.

---

*[pkg/ccl/sqlccl/backup_test.go, line 288 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kg_rO78dqNHBqF5COV7:-KgaND0e00ppOpYtov0e:b-aigaa5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4cbb9ad02ea8404c20d20ed11c2f87819b02e98/pkg/ccl/sqlccl/backup_test.go#L288)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

an entire sentence in parentheses?
</blockquote></details>

FWIW http://english.stackexchange.com/a/78174

(But sure, done.)

---

*[pkg/sql/jobs_test.go, line 116 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kg_oMwUM_njyT_oOSAb:-KgaNPo-fnEDQs0rgGns:b-x4mtzo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e66161a4ec2ae4a7b4e292e78ec1d6bc716276/pkg/sql/jobs_test.go#L116)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

i think this might be clearer as a `switch status`?
</blockquote></details>

Hmmm...

```go
if expected.Before.After(created.Time) {
    return errors.Errorf(
        "created time %v is before expected created time %v",
        created, expected.Before,
    )
}

if started.Valid && created.Time.After(started.Time) {
    return errors.Errorf("created time %v is after started time %v", created, started)
}

switch (status) {
case sql.JobStatusPending:
  return verifyModifiedAgainst("created", created.Time)
case sql.JobStatusRunning:
  return verifyModifiedAgainst("started", started.Time)
case sql.JobStatusSucceeded:
  if !started.Valid {
    return errors.Errorf("started time is NULL but job claims to be successful")
  }
  fallthrough  
case sql.JobStatusFailed:
  if started.Time.After(finished.Time) {
    return errors.Errorf("started time %v is after finished time %v", started, finished)
  }
  if e, a := expected.Error, err; e != a {
    return errors.Errorf("expected error %v, got %v", e, a)
  }
  return verifyModifiedAgainst("finished", finished.Time)
}
```

Thoughts? It's a little clunky no matter how you write it. (If you feel strongly I'll address in a separate PR since I don't want to further expand the scope of this one.)

---

*[pkg/sql/jobs_test.go, line 350 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kg_opHizBvCO2ICurmR:-KgaPVvi8Qkhk627MOSL:b-6s2fso) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e66161a4ec2ae4a7b4e292e78ec1d6bc716276/pkg/sql/jobs_test.go#L350)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

suggest you run this with `-v` before you settle on multiword subtest names
</blockquote></details>

This seems fine to me:

```
=== RUN   TestJobLogger/valid_job_lifecycles_succeed
=== RUN   TestJobLogger/bad_job_details_fail
=== RUN   TestJobLogger/update_before_create_fails
=== RUN   TestJobLogger/same_state_transition_twice_fails
=== RUN   TestJobLogger/out_of_bounds_progress_fails
=== RUN   TestJobLogger/progress_on_non-started_job_fails
=== RUN   TestJobLogger/progress_on_finished_job_fails
=== RUN   TestJobLogger/succeeded_forces_fraction_completed_to_1.0
```

do you have a preferred alternative?

---

*[pkg/sql/jobs_test.go, line 178 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kg_p3O-l8NwlAhU2NR4:-Kglub3bYo4hRgkaLh-z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ebd9bd67caa2c7d64af35aa82ded0d6584ee94b/pkg/sql/jobs_test.go#L178)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this belongs in the preceding commit
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r11.
Review status: 4 of 5 files reviewed at latest revision, 21 unresolved discussions, all commit checks successful.

---

*[pkg/sql/jobs_test.go, line 116 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kg_oMwUM_njyT_oOSAb:-KgoOH43L9F66blj8iQR:b-bhazff) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e66161a4ec2ae4a7b4e292e78ec1d6bc716276/pkg/sql/jobs_test.go#L116)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Hmmm...

```go
if expected.Before.After(created.Time) {
    return errors.Errorf(
        "created time %v is before expected created time %v",
        created, expected.Before,
    )
}

if started.Valid && created.Time.After(started.Time) {
    return errors.Errorf("created time %v is after started time %v", created, started)
}

switch (status) {
case sql.JobStatusPending:
  return verifyModifiedAgainst("created", created.Time)
case sql.JobStatusRunning:
  return verifyModifiedAgainst("started", started.Time)
case sql.JobStatusSucceeded:
  if !started.Valid {
    return errors.Errorf("started time is NULL but job claims to be successful")
  }
  fallthrough  
case sql.JobStatusFailed:
  if started.Time.After(finished.Time) {
    return errors.Errorf("started time %v is after finished time %v", started, finished)
  }
  if e, a := expected.Error, err; e != a {
    return errors.Errorf("expected error %v, got %v", e, a)
  }
  return verifyModifiedAgainst("finished", finished.Time)
}
```

Thoughts? It's a little clunky no matter how you write it. (If you feel strongly I'll address in a separate PR since I don't want to further expand the scope of this one.)
</blockquote></details>

That fallthrough is not actually legal, unfortunately. I believe `fallthrough` has to be by itself in a `case`. I do think this is better though.

---

*[pkg/sql/jobs_test.go, line 350 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kg_opHizBvCO2ICurmR:-KgoOW_C5pfpj-NuY-88:bxkbena) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e66161a4ec2ae4a7b4e292e78ec1d6bc716276/pkg/sql/jobs_test.go#L350)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

also seems fine to me
</blockquote></details>

Seems ok.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 2 of 5 files reviewed at latest revision, 21 unresolved discussions.

---

*[pkg/ccl/sqlccl/backup.go, line 388 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kg_q9ZRvPxO8nqch0XH:-KgobKAuU4GRTWM6IROL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ebd9bd67caa2c7d64af35aa82ded0d6584ee94b/pkg/ccl/sqlccl/backup.go#L388)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

The duplication here bothers me, too. my only concern with two mutexes is deadlock but it should be easy to avoid that here

I'd extract the BackupRestore one now, but with a TODO that it could generalize well to schema changes
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup_test.go, line 340 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgVc3wOMV6cS15PaWcS:-KgokizUBkze0Wd4bOCt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68696fd555bde73e3086d7aa3e93737cc111813/pkg/ccl/sqlccl/backup_test.go#L340)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

The test will timeout if it hangs forever, so not worth adding code complexity to avoid that.

If we got back to only testing it halfway through and checking that it's between 25% and 75%, I think that's likely to pass even without the RetryForDuration (though we should keep it to guard against flakes). Plus it's likely the test will run faster, which is nice when stressing it. I'm just not convinced the assertion that we see N distinct values is buying us much
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup_test.go, line 395 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgoGHXpgwphqKx9Zka0:-KgoNZJbkTIofcvjzTDN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/80ce47d60d7ef34dad6aaafb2dd62e102e633580/pkg/ccl/sqlccl/backup_test.go#L395)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

two lines plz
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 3 files at r13.
Review status: 4 of 5 files reviewed at latest revision, 21 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/backup.go, line 58 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kgol17VAmrjP6LRxPqd:-Kgol17VAmrjP6LRxPqe:bxfrcvl) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ccffd366614937917c620be2091990a84b4a9/pkg/ccl/sqlccl/backup.go#L58)):*
> ```Go
> 
> type jobProgressLogger struct {
> 	// Fields to initialize
> ```

period at the end of sentences other than in inline comments.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 4 of 5 files reviewed at latest revision, 21 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/backup.go, line 58 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kgol17VAmrjP6LRxPqd:-KgolLEfO7CYLEW8mUo4:b-2uf1lv) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ccffd366614937917c620be2091990a84b4a9/pkg/ccl/sqlccl/backup.go#L58)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

period at the end of sentences other than in inline comments.
</blockquote></details>

Well, they're technically not sentences since they don't have a verb. Should I make them sentences or just slap a period on the end?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459)*
<!-- Sent from Reviewable.io -->

tamird commented:
You're right; my comment should have been "comments that aren't inline
fragments should be proper sentences with punctuation.".

On Mon, Apr 3, 2017 at 1:04 PM, Nikhil Benesch <notifications@github.com>
wrote:

> Review status: 4 of 5 files reviewed at latest revision, 21 unresolved
> discussions, some commit checks pending.
> ------------------------------
>
> *pkg/ccl/sqlccl/backup.go, line 58 at r13
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kgol17VAmrjP6LRxPqd:-KgolLEfO7CYLEW8mUo4:b-2uf1lv>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/989ccffd366614937917c620be2091990a84b4a9/pkg/ccl/sqlccl/backup.go#L58>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> period at the end of sentences other than in inline comments.
>
> Well, they're technically not sentences since they don't have a verb.
> Should I make them sentences or just slap a period on the end?
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14459>*
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/14459#issuecomment-291206955>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPBBDc1JV3RWYhfHRIHjWX472HYIuks5rsSaTgaJpZM4Mto9M>
> .
>

benesch commented:
Done.

---

Review status: 3 of 5 files reviewed at latest revision, 21 unresolved discussions.

---

*[pkg/ccl/sqlccl/backup.go, line 58 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kgol17VAmrjP6LRxPqd:-Kgon1KmwfIQbJYeml9h:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ccffd366614937917c620be2091990a84b4a9/pkg/ccl/sqlccl/backup.go#L58)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Well, they're technically not sentences since they don't have a verb. Should I make them sentences or just slap a period on the end?
</blockquote></details>

Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-:-Kgon72nTQcR99gDY7ON:b-896fix)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r14.
Review status: 4 of 5 files reviewed at latest revision, 20 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: 4 of 5 files reviewed at latest revision, 26 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/backup.go, line 63 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgomzVgwStPThB3oYeI:-KgomzVh6CBQakO_6w26:b-gpuu7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ccffd366614937917c620be2091990a84b4a9/pkg/ccl/sqlccl/backup.go#L63)):*
> ```Go
> 
> 	// Internal fields
> 	syncutil.Mutex
> ```

I like the anonymous embedded struct patter we use elsewhere for this

```
type jobProgressLogger struct {
  ...

  mu struct {
    syncutil.Mutex
    chunkCount int
    ...
  }
}
```

Then accesses to protected fields look like `jobProgress.mu.chunkCount` which makes it real obvious which locks you need to hold

---

*[pkg/ccl/sqlccl/backup.go, line 64 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgonnNEdHB5QXxz-tzv:-KgonnNEdHB5QXxz-tzw:b1mkasy) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ccffd366614937917c620be2091990a84b4a9/pkg/ccl/sqlccl/backup.go#L64)):*
> ```Go
> 	// Internal fields
> 	syncutil.Mutex
> 	chunkCount           int
> ```

nit: completedChunks (to parallel totalChunks)

---

*[pkg/ccl/sqlccl/backup.go, line 82 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgonVFdIPNeCDrtHS_V:-KgonVFdIPNeCDrtHS_W:b-k542kb) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ccffd366614937917c620be2091990a84b4a9/pkg/ccl/sqlccl/backup.go#L82)):*
> ```Go
> }
> 
> func (jpl *jobProgressLogger) shouldLogProgress(fraction float32) bool {
> ```

comment that this needs to be called while under mutex. better yet, inline it

---

*[pkg/ccl/sqlccl/backup_test.go, line 296 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kgoo36_q0ImjMi53Qxg:-Kgoo36a1n-28MHXgwaf:blqf076) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ccffd366614937917c620be2091990a84b4a9/pkg/ccl/sqlccl/backup_test.go#L296)):*
> ```Go
> ) error {
> 	<-responseComplete
> 	err := util.RetryForDuration(time.Second, func() error {
> ```

use testutils.SucceedsSoon (unless there's some particular reason it needs to be 1s instead of DefaultSucceedsSoonDuration). don't forget to update the "1s" in your function level comment

---

*[pkg/ccl/sqlccl/backup_test.go, line 319 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgooNAsCA2nMjUA9wPY:-KgooNAtwe4AMOXM0XXK:bv33ysb) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ccffd366614937917c620be2091990a84b4a9/pkg/ccl/sqlccl/backup_test.go#L319)):*
> ```Go
> 	// This test expects a progress update for at least each range that we've
> 	// backed up, which requires a sufficiently small progressFractionThreshold.
> 	if 1.0/float32(backupRestoreDefaultRanges) < progressFractionThreshold {
> ```

this isn't necessary anymore, right?

---

*[pkg/ccl/sqlccl/backup_test.go, line 354 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgopuaLasNgU4KbqJZ3:-KgopuaMDq-JX8iw5iHA:b2x6hd2) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ccffd366614937917c620be2091990a84b4a9/pkg/ccl/sqlccl/backup_test.go#L354)):*
> ```Go
> 			mu.Lock()
> 			defer mu.Unlock()
> 			if mu.responseCompleteCount == expectedProgressUpdateCount/2 {
> ```

as discussed offline, I think this could be simplified

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 3 of 6 files reviewed at latest revision, 26 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/backup.go, line 63 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgomzVgwStPThB3oYeI:-Kgp-w_YCCv62YzRIpTw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ccffd366614937917c620be2091990a84b4a9/pkg/ccl/sqlccl/backup.go#L63)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I like the anonymous embedded struct patter we use elsewhere for this

```
type jobProgressLogger struct {
  ...

  mu struct {
    syncutil.Mutex
    chunkCount int
    ...
  }
}
```

Then accesses to protected fields look like `jobProgress.mu.chunkCount` which makes it real obvious which locks you need to hold
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.go, line 64 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgonnNEdHB5QXxz-tzv:-Kgp-wHNjE3EFvIllRSE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ccffd366614937917c620be2091990a84b4a9/pkg/ccl/sqlccl/backup.go#L64)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: completedChunks (to parallel totalChunks)
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.go, line 82 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgonVFdIPNeCDrtHS_V:-Kgp-xFkA400IOq58-tB:b8wd1c1) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ccffd366614937917c620be2091990a84b4a9/pkg/ccl/sqlccl/backup.go#L82)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

comment that this needs to be called while under mutex. better yet, inline it
</blockquote></details>

Inlined. My gripe is that `go fmt` thinks this is a reasonable format for multi-line conditionals

```go
if fraction-jpl.mu.lastReportedFraction > progressFractionThreshold ||
  jpl.mu.lastReportedAt.Add(progressTimeThreshold).Before(timeutil.Now()) {
  return jpl.jobLogger.Progressed()
}
```

but it's nearly impossible to tell where the condition starts and the code begins. I can live with it, though, especially given the separating comment in this case.

---

*[pkg/ccl/sqlccl/backup_test.go, line 296 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kgoo36_q0ImjMi53Qxg:-Kgp0W9rjqdF3-PfLSAD:b76ctju) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ccffd366614937917c620be2091990a84b4a9/pkg/ccl/sqlccl/backup_test.go#L296)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

use testutils.SucceedsSoon (unless there's some particular reason it needs to be 1s instead of DefaultSucceedsSoonDuration). don't forget to update the "1s" in your function level comment
</blockquote></details>

`SucceedsSoon` takes a `testing.TB`, and I was trying to avoid passing that in to this helper. Thoughts? Which is the lesser evil?

---

*[pkg/ccl/sqlccl/backup_test.go, line 319 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgooNAsCA2nMjUA9wPY:-Kgp0aUHtB-fz8kDTz4-:bgjn3sx) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ccffd366614937917c620be2091990a84b4a9/pkg/ccl/sqlccl/backup_test.go#L319)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

this isn't necessary anymore, right?
</blockquote></details>

It's *less* necessary, because we only need `progressFractionThreshold <= 0.5`.  I've killed it because the chance that we ever set `progressFractionThreshold` to something greater than 0.5 is nil.

---

*[pkg/ccl/sqlccl/backup_test.go, line 354 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-KgopuaLasNgU4KbqJZ3:-Kgp1P5Qv0H49n7W_OHt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ccffd366614937917c620be2091990a84b4a9/pkg/ccl/sqlccl/backup_test.go#L354)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

as discussed offline, I think this could be simplified
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459)*
<!-- Sent from Reviewable.io -->

danhhz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 3 of 6 files reviewed at latest revision, 28 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/backup_test.go, line 296 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kgoo36_q0ImjMi53Qxg:-Kgp2vsDkipepSLlvLN9:b5hf6hl) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ccffd366614937917c620be2091990a84b4a9/pkg/ccl/sqlccl/backup_test.go#L296)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

`SucceedsSoon` takes a `testing.TB`, and I was trying to avoid passing that in to this helper. Thoughts? Which is the lesser evil?
</blockquote></details>

gotcha. this is fine.

---

*[pkg/ccl/sqlccl/progress.go, line 1 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kgp2YeC7ZoImL6gsOQ4:-Kgp2YeC7ZoImL6gsOQ5:bb6ht2h) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f1ed26bea63c8a1fea0bc4bc216104c0a7ba822/pkg/ccl/sqlccl/progress.go#L1)):*
> ```Go
> // Copyright 2016 The Cockroach Authors.
> ```

2017

---

*[pkg/ccl/sqlccl/progress.go, line 52 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-Kgp2dGJUd6rPRkvnejz:-Kgp2dGKG6AI-4TqK5iL:betl29w) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f1ed26bea63c8a1fea0bc4bc216104c0a7ba822/pkg/ccl/sqlccl/progress.go#L52)):*
> ```Go
> 	jpl.mu.completedChunks++
> 	fraction := float32(jpl.mu.completedChunks) / float32(jpl.totalChunks)
> 	if fraction-jpl.mu.lastReportedFraction > progressFractionThreshold ||
> ```

maybe name this bool, though: `shouldUpdateProgress := ...; if shouldUpdateProgress { ... }`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459#-:-Kgp31dNtj7JesysDs5C:bnfp4nl)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Reviewed 1 of 4 files at r7, 1 of 3 files at r11, 1 of 3 files at r13, 2 of 3 files at r15, 1 of 1 files at r16.
Review status: all files reviewed at latest revision, 28 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14459)*
<!-- Sent from Reviewable.io -->
